### PR TITLE
util/syspolicy/source: add package for reading policy settings from external stores

### PR DESCRIPTION
**Pull request by nickkhyl**
We add package defining interfaces for policy stores, enabling creation of policy sources and reading settings from them. It includes a Windows-specific `PlatformPolicyStore` for GP and MDM policies stored in the Registry, and an in-memory `TestStore` for testing purposes.

We also include an internal package that tracks and reports policy usage metrics when a policy setting is read from a store. Initially, it will be used only on Windows and Android, as macOS, iOS, and tvOS report their own metrics. However, we plan to use it across all platforms eventually.

Updates #12687

### CODE REVIEW

**Review by sailorfrag — COMMENTED** [review]
Still need to do `policy_source.go`, `policy_store_windows.go` and `test_store.go`, along with associated tests. But here are comments so far.

File: util/syspolicy/internal/metrics/metrics.go
Comment (nickkhyl) [reply util/syspolicy/internal/metrics/metrics.go:32]: We plan to use it across all platforms, including iOS/macOS, which currently report their own metrics, and Unix-like platforms, which lack `syspolicy` support today. At a minimum, I think we can implement an [envvar-based policy store](https://github.com/tailscale/tailscale/issues/13193#issuecomment-2298084842) and/or a config-like policy store. Additionally, there are discussions about allowing `syspolicy` configuration via the control plane, which would serve as a cross-platform policy source.

That being said, I definitely want to exclude test-related code from production builds, especially on iOS. My original plan was to move certain types and functions to `_notios.go` files in a follow-up PR.
However, I'm very interested in the following option and wasn't aware that our CI supports it.
> For testing, we could also use a build tag to include the full version for tests.

Do you know of a good example where we use this approach?

File: util/syspolicy/internal/metrics/metrics.go
Comment (nickkhyl) [reply util/syspolicy/internal/metrics/metrics.go:79]: Using a counter, we'll report the number of times a resultant policy is changed multiplied by the total number of errored policy settings in the applicable policy stores.

For example, if there are two policy settings with invalid values and we re-read the policy 10 times, we'll report 20 new errors. The number of times we've re-read the policy within an hour, day, or week is out of our control and is heavily influenced by the environment and user behavior. Therefore, I'm struggling to understand why a counter is more appropriate for the `windows_syspolicy_errors` metric, especially since all other `syspolicy` metrics are gauges.

File: util/syspolicy/internal/metrics/metrics.go
Comment (nickkhyl) [reply util/syspolicy/internal/metrics/metrics.go]: I think we should add a `tailscale debug` subcommand and log all configured policies when generating a bug report. I believe this is already supported for Windows and Registry-based policy settings. However, it would be ideal to log the actual resultant policy merged from all policy sources and have this functionality across all platforms.

Would it be okay to keep logging in addition to that?

File: util/syspolicy/internal/metrics/metrics.go
Comment (nickkhyl) [reply util/syspolicy/internal/metrics/metrics.go]: whoops. I'm unsure why I made it a type alias. Fixed. Thanks!

File: util/syspolicy/internal/metrics/metrics.go
Comment (nickkhyl) [reply util/syspolicy/internal/metrics/metrics.go]: I originally considered including the leading underscore in the `metricScopeName`, but didn't like it at another callsite.

But yeah, I think the `strings.Join` approach is better. So I updated `newMetric` to accept a slice of components and use `slicesx.Filter` + `strings.Join` to produce the actual metric name from parts.
Thanks!

File: util/syspolicy/source/policy_reader.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_reader.go]: It was the type name I originally chose for what later became a `Reader`.
It wasn't in square brackets, so I guess VSCode didn't pick it up when I renamed the type.
Thanks for noticing!

File: util/syspolicy/source/policy_reader.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_reader.go:44]: It is small enough when there's just one `setting.Origin`, but it all adds up (and rounds up due to alignment) when used in  each `policy.Item` / `policy.List` within a struct like this:
```go
// ExitNodes is a policy group that controls utilizing and running exit nodes.
type ExitNodes struct {
	// AllowedSuggestedExitNodes is a list of exit node IDs that restricts which
	// exit nodes are considered when generating suggestions for exit nodes.
	AllowedSuggestedExitNodes policy.List[string] ` policy:",ios,macos,windows"`
	// AllowLANAccess controls whether whether locally accessible
	// subnets should be routed directly or via the exit node.
	AllowLANAccess policy.Item[PreferenceOption] `policy:"/ExitNodeAllowLANAccess"`
	// ExitNodeID is the exit node's node id.
	// ExitNodeID takes precedence over ExitNodeIP.
	// To find the node ID, go to /api.md#device.
	ExitNodeID policy.Item[string]
	// ExitNodeIP is the exit node's IP address.
	ExitNodeIP policy.Item[string]
	// RunExitNode controls if the device acts as an exit node. Even when
	// running as an exit node, the device must be approved by a tailnet
	// administrator. Its name is slightly awkward because RunExitNodeVisibility
	// predates this option but is preserved for backwards compatibility.
	RunExitNode policy.Item[PreferenceOption] `policy:"/AdvertiseExitNode"`
}
```
I'll submit the `syspolicy/policy` package and the actual policy structures in a future PR once this and the next one merge.

But my reasoning was to maintain a single instance of `setting.Origin` for all policy settings with the same origin, and include it in the `policy.Item` and `policy.List` structs by pointer rather than by value.
We can revisit this decision in the future PR if you don't mind.

File: util/syspolicy/source/policy_reader.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_reader.go]: Ah, right. [Logger.Output](https://pkg.go.dev/log#Logger.Output):
> Output writes the output for a logging event. The string s contains the text to print after the prefix specified by the flags of the Logger. **_A newline is appended if the last character of s is not already a newline_.**
Thanks!

File: util/syspolicy/source/policy_reader.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_reader.go]: Ouch. It surely does. Thanks for catching that!

File: util/syspolicy/source/policy_reader.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_reader.go]: gotcha; changed to a plain mutex. Thanks!

File: util/syspolicy/source/policy_reader.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_reader.go]: nope. We should always add a `RawItem` to the map. However, the error (and hence the error text in the `RawItem`) may be nil or non-nil.

I changed that line to (hopefully) make it easier to read.

File: util/syspolicy/source/policy_reader.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_reader.go:311]: The sources slice may contain multiple sources with the same scope. This method returns the index of a specific `*Source` slice element rather than the index of any source with the same scope as the specified source.
So no, we cannot use binary search here. Additionally, the sources slice is short enough for the linear search to probably work faster.

File: util/syspolicy/source/policy_reader.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_reader.go:317]: Nope. I'd like to insert new sources after any existing sources with the same scope, and this method is essentially an `std::upper_bound`. I couldn't find an existing implementation of it in Go's stdlib.
Additionally, `slices.BinarySearch` requires elements to be sorted in the strictly increasing order, but the readable sources are actually sorted in non-decreasing order.

I should probably clarify all that in a docstring.
Alternatively, maybe we should change this to a good ol' linear search?
There's likely no real benefit in using binary search on a slice of just a few policy sources, and perhaps a loop would be less confusing than a hand-rolled binary search.

File: util/syspolicy/source/policy_reader.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_reader.go:347]: Almost never. We might have to change the list of sources when switching between Tailscale profiles. We could also choose to do it in `tailscale-ipn` when it connects to the backend, which would occur once per OS user login.
Additionally, it's hard to imagine more than 3-7 policy sources, and each `ReadableSource` is just two words.

File: util/syspolicy/source/policy_reader.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_reader.go:115]: yes. I'm just unsure if relaxing it is a good idea; I chose not to.
If the caller wants to actually read the policy from the underlying store and calls the `ReadSetting` method, I think we shouldn't be returning an outdated policy snapshot we read last week.

File: util/syspolicy/internal/metrics/metrics.go
Comment (sailorfrag) [reply util/syspolicy/internal/metrics/metrics.go:32]: no, but you just have to do something like `//go:build windows || android || test` -- the convention of putting tests in `foo_test.go` is using the compiler's implicit filename-based build tag logic to avoid including the test files if not running a test.

Comment (sailorfrag) [reply util/syspolicy/internal/metrics/metrics.go:32]: Anyway to address the bigger response about planning to use it for all platforms, that's fine, but the current iteration isn't for all platforms, so it feels like it might be nice to avoid including it non the currently non-supported ones? If the plan isn't for things to stay that way, then don't put a ton of effort into it, but I don't think it should be super hard.

Anyway, assuming it's not putting us over any size thresholds we're worried about, I guess it's an optional suggestion.

Comment (nickkhyl) [reply util/syspolicy/internal/metrics/metrics.go:32]: I see. But the `_test` suffix works differently and does not set the `test` build tag.
Instead, it makes the test code a part of a test package.

So this:
```
//go:build windows || android || test
```
will work if you run the tests on Windows (or Android if that's possible).
But it won't work when running tests on Linux unless we _explicitly_ set the `test` tag when we run the tests.
Hence the question about the CI support.

Comment (nickkhyl) [reply util/syspolicy/internal/metrics/metrics.go:32]: That said, I agree with your second point and am happy to hide what's possible under the `!ios` tag (or its file suffix equivalent).

File: util/syspolicy/internal/metrics/metrics.go
Comment (sailorfrag) [reply util/syspolicy/internal/metrics/metrics.go]: Yeah, I think a `tailscale debug component-logs` category would be useful to debug problems loading policies from the store, and bugreport + `tailscale debug` subcommand to dump the merged policy. They have different uses. The former is more about making it easier for `loggerx.Verbosef` to actually be enabled.

File: util/syspolicy/source/policy_reader.go
Comment (sailorfrag) [reply util/syspolicy/source/policy_reader.go:115]:
```golang
// ReadSettings reads policy settings from the underlying [Store] even if no
// changes were detected. It returns the new [*setting.Snapshot],nil on
// success or an undefined snapshot (possibly `nil`) along with a non-`nil`
// error in case of failure.
```
😁 

I leave it up to you, it's fine as-is.

File: util/syspolicy/internal/metrics/metrics.go
Comment (nickkhyl) [reply util/syspolicy/internal/metrics/metrics.go:79]: Yeah, now I'm unsure of the value in having this specific metric at all 😁 

I think my overall feeling is that the "number of policies that failed to load" is less useless than the "number of times an error occurred", even without knowing the denominator for the first one. The second metric could be greatly skewed by a customer with ephemeral nodes or a customer who has an unusually short Group Policy processing interval (e.g., 1 minute vs. the standard 90 minutes), and so on.

I think I'll keep it as is for now with the TODO in place, but add a link to this thread from the code.

Also, maybe @kabirsikand has any insights on how the `windows_syspolicy_errors` and `macos_syspolicy_errors`/`ios_syspolicy_errors` metrics are used?

File: util/syspolicy/source/policy_store_windows.go
Comment (sailorfrag) [inline util/syspolicy/source/policy_store_windows.go:82]: Does it need to be stated that the token will be duplicated?

File: util/syspolicy/source/policy_store_windows.go
Comment (sailorfrag) [inline util/syspolicy/source/policy_store_windows.go:123]: Maybe the policy lock should be a parameter, populated by the caller? The caller is already specialized to machine vs user.

File: util/syspolicy/source/policy_store_windows.go
Comment (sailorfrag) [inline util/syspolicy/source/policy_store_windows.go:162]: Should it trigger a close if it's going to return `ErrStoreClosed`?

File: util/syspolicy/source/policy_store_windows.go
Comment (sailorfrag) [inline util/syspolicy/source/policy_store_windows.go:241]: The lock is going to be held so infrequently that I think it makes more sense to use a normal `Mutex` rather than `RWMutex`. I don't think it matters if it's possible to have two `onChange()` instances running in parallel or not.

File: util/syspolicy/source/policy_store_windows.go
Comment (sailorfrag) [inline util/syspolicy/source/policy_store_windows.go:296]: Hmm, is that because `REG_MULTI_SZ` is a recent addition and there's been a need for such a thing for a long time? Just curious.

That's annoying but thanks for the attention to detail here.

File: util/syspolicy/source/policy_store_windows.go
Comment (sailorfrag) [inline util/syspolicy/source/policy_store_windows.go:329]: I guess this is okay. I feel like it might end up a little more logical looking if you split on the separator and then join with backslash, but I don't think one way is significantly better than the other.

File: util/syspolicy/source/policy_store_windows.go
Comment (sailorfrag) [inline util/syspolicy/source/policy_store_windows.go:324]: Maybe describe what's going on in a docstring.

File: util/syspolicy/source/policy_store_windows.go
Comment (sailorfrag) [inline util/syspolicy/source/policy_store_windows.go:377]: It's preferred not to defer in a loop. It would be easy enough to just explicitly close after calling `getValue()` before checking `err`.

File: util/syspolicy/source/policy_store_windows.go
Comment (sailorfrag) [inline util/syspolicy/source/policy_store_windows.go:403]: Maybe move this after the check for `ps.closing` because a double close will cause a panic.

File: util/syspolicy/source/policy_store_windows_test.go
Comment (sailorfrag) [inline util/syspolicy/source/policy_store_windows_test.go:37]: Might be good to make sure the test is verifying all the goroutines are gone at the end of the test and can't leak.

File: util/syspolicy/internal/metrics/metrics.go
Comment (nickkhyl) [reply util/syspolicy/internal/metrics/metrics.go]: Gotcha. I'll add a `SetDebugLoggingEnabled` function and update CLI and `LocalBackend` to support it, but would prefer to do it in its own PR.

https://github.com/tailscale/tailscale/issues/13313

File: util/syspolicy/source/policy_reader.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_reader.go:115]: If behavior like this is common and expected for Go funcs then.. sure.

I updated it and the docstring.

File: util/syspolicy/source/policy_source.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_source.go:55]: I did consider it. It's just very unfortunate it's not allowed to fail.

File: util/syspolicy/source/policy_source.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_source.go:116]: It appears we're not always that strict about it. But yeah, in most cases this is accurate:
> `fmt.Stringer` is only applicable if it can fully convey the contents of the struct, rather than a summary of it.
So let's stick to it.

Renamed to `Description()`, thanks!

File: util/syspolicy/source/policy_source.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_source.go:130]: Maybe. I've been considering something similar to [allocating an altitude](https://learn.microsoft.com/en-us/windows-hardware/drivers/ifs/allocated-altitudes) in the future when we start adding more policy sources, or just have an enum where each supported kind of a policy source is assigned a specific number.

I think I'd like like to keep the insertion as is, and if multiple sources have the same precedence values (regardless of how we determine it), then the recently added policy sources should take precedence over the older ones.

File: util/syspolicy/source/policy_store_windows.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_store_windows.go:92]: I generally agree. But we don't have such a function and use concatenation throughout the codebase, and in this particular case we're concatenating a const that doesn't end with a `\` to a SID string (which never starts with a `\`).
So I felt like adding a `Join`-like function here is a bit redundant.

I'll keep it as is for now.

File: util/syspolicy/source/policy_store_windows.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_store_windows.go:82]: It shouldn't matter for the caller that the token will be duplicated, but I think it is worth mentioning that `[windows.TOKEN_DUPLICATE]` is also required.

I updated the docstring. Thanks!

File: util/syspolicy/source/policy_store_windows.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_store_windows.go:162]: No. It actually means that the close is already in progress and we've lost the race.
I added a comment.

File: util/syspolicy/source/policy_store_windows.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_store_windows.go:241]: We also acquire it in `getPolicyValue` when reading any policy setting from the source.
But yeah.. I agree that we probably don't need an RW lock in `PlatformPolicyStore` given the downsides you've mentioned the other day.

Changed to a plain `Mutex`.

File: util/syspolicy/source/policy_store_windows.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_store_windows.go:296]: `REG_MULTI_SZ` was added back in the nineties. I'm not perfectly sure when exactly it became available, but it certainly pre-dates the Group Policy in general (added in NT 4.0), and ADMX specifically (added in Vista).

I'd say the main reasons why Group Policy / Administrative Templates don't use it for lists are:
- Mergeability (additivity?) of changes made in different GPOs applied to different OUs. That is, in certain cases if the same policy setting is configured in more than one place, the resultant "value" observed on the endpoints includes all the added items. And this doesn't require any special handling; the GP engine just writes values, one by one, as it reads the `Registry.pol` file from each applicable GPO.
- It's easier for to check policy settings this way, both in the user and kernel mode. Since the Name and the Value are the same for list items, an app or a driver could just check if a registry value with a given name exists if it needs to check if, say, a certain domain is okay to visit, is a member of a network zone, and so on. Otherwise, it would either have to re-read and check the entire list of values every time, or keep an in-memory copy of the most recent values as we're about to do. In most cases, it's much easier (and arguably faster in a lot of cases) to just check whether a registry value exists.
- It's compatible with another type of policy setting, which is essentially a string map. It's similar to the list setting type, but unlike the list, an admin can also configure a value for a given key. But it's the same code under the hood and requires no special handling on the admin console, GPClient service, or app side. If an app evolves, it can probably even transition from a list to a dictionary, and maintain compatibility with any existing policies.

File: util/syspolicy/source/policy_store_windows.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_store_windows.go:329]: I would expect less (re-)allocations in the current implementations.

File: util/syspolicy/source/policy_store_windows.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_store_windows.go:324]: Done.

File: util/syspolicy/source/policy_store_windows.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_store_windows.go:403]: Good point. I think sometime earlier the ordering was important, but then I fixed the underlying issue and forgot to move this after the `ps.closing` check.

File: util/syspolicy/source/policy_store_windows_test.go
Comment (nickkhyl) [reply util/syspolicy/source/policy_store_windows_test.go:37]: Added
```go
// Make sure we don't leak goroutines
tstest.ResourceCheck(t)
```

**Review by sailorfrag — APPROVED** [review]
Thanks for bearing with me.

**Pull request merged**
